SMTP office365

### 检查清单

- [X] 已经在 [GitHub Issues](https://github.com/halo-dev/halo/issues) 中搜索过该问题
- [X] 已经在[论坛](https://bbs.halo.run/)中搜索过该问题
- [X] 已经查阅过[文档](https://docs.halo.run/)

### 是什么版本出现了此问题？

1.5.2

### 使用的什么数据库？

MySQL 8.x

### 使用的哪种方式部署？

Fat Jar

### 在线站点地址

https://www.wangxiaoer.site

### 发生了什么？

邮件发送 smtp设置为office365 连接会报错


### 相关日志输出

```shell
Apr 16 02:48:46 instance-d java[7058]: Exception in thread "pool-4-thread-2" run.halo.app.exception.EmailException: 邮件发送失败，请检查 SMTP 服务配置是否正确                                                    
Apr 16 02:48:46 instance-d java[7058]:         at run.halo.app.mail.AbstractMailService.sendMailTemplate(AbstractMailService.java:122)      
Apr 16 02:48:46 instance-d java[7058]:         at run.halo.app.mail.AbstractMailService.lambda$sendMailTemplate$0(AbstractMailService.java:137)                                                                   
Apr 16 02:48:46 instance-d java[7058]:         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) 
Apr 16 02:48:46 instance-d java[7058]:         
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) Apr 16 02:48:46 instance-d java[7058]:         
at java.base/java.lang.Thread.run(Thread.java:833)                                           
Apr 16 02:48:46 instance-d java[7058]: Caused by: org.springframework.mail.MailSendException: Mail server connection failed; nested exception is javax.mail.MessagingException: Could not connect to SMTP host: smtp.office365.com, port: 587;                                          
Apr 16 02:48:46 instance-d java[7058]:   nested exception is:         
Apr 16 02:48:46 instance-d java[7058]:         javax.net.ssl.SSLException: Unsupported or unrecognized SSL message. Failed messages: javax.mail.MessagingException: Could not connect to SMTP host: smtp.office365.com, port: 587;                                                      
Apr 16 02:48:46 instance-d java[7058]:   nested exception is:         
Apr 16 02:48:46 instance-d java[7058]:         javax.net.ssl.SSLException: Unsupported or unrecognized SSL message; message exceptions (1) are:                                                                   
Apr 16 02:48:46 instance-d java[7058]: Failed message 1: javax.mail.MessagingException: Could not connect to SMTP host: smtp.office365.com, port: 587;                                                            
Apr 16 02:48:46 instance-d java[7058]:   nested exception is:         
Apr 16 02:48:46 instance-d java[7058]:         javax.net.ssl.SSLException: Unsupported or unrecognized SSL message                          
Apr 16 02:48:46 instance-d java[7058]:         at org.springframework.mail.javamail.JavaMailSenderImpl.doSend(JavaMailSenderImpl.java:448)  
Apr 16 02:48:46 instance-d java[7058]:         at org.springframework.mail.javamail.JavaMailSenderImpl.send(JavaMailSenderImpl.java:361)    
Apr 16 02:48:46 instance-d java[7058]:         at org.springframework.mail.javamail.JavaMailSenderImpl.send(JavaMailSenderImpl.java:356)    
Apr 16 02:48:46 instance-d java[7058]:         at run.halo.app.mail.AbstractMailService.sendMailTemplate(AbstractMailService.java:115)      
Apr 16 02:48:46 instance-d java[7058]:         ... 4 more             Apr 16 02:48:46 instance-d java[7058]: Caused by: javax.mail.MessagingException: Could not connect to SMTP host: smtp.office365.com, port: 587;                                                                   
Apr 16 02:48:46 instance-d java[7058]:   nested exception is:         Apr 16 02:48:46 instance-d java[7058]:         javax.net.ssl.SSLException: Unsupported or unrecognized SSL message                          
Apr 16 02:48:46 instance-d java[7058]:         at com.sun.mail.smtp.SMTPTransport.openServer(SMTPTransport.java:2212)                       
Apr 16 02:48:46 instance-d java[7058]:         at com.sun.mail.smtp.SMTPTransport.protocolConnect(SMTPTransport.java:722)                   
Apr 16 02:48:46 instance-d java[7058]:         at javax.mail.Service.connect(Service.java:342)                                              
Apr 16 02:48:46 instance-d java[7058]:         at org.springframework.mail.javamail.JavaMailSenderImpl.connectTransport(JavaMailSenderImpl.java:518)                                                              
Apr 16 02:48:46 instance-d java[7058]:         at org.springframework.mail.javamail.JavaMailSenderImpl.doSend(JavaMailSenderImpl.java:437)  Apr 16 02:48:46 instance-d java[7058]:        
 ... 7 more             
 Apr 16 02:48:46 instance-d java[7058]: Caused by: javax.net.ssl.SSLException: Unsupported or unrecognized SSL message                       
 Apr 16 02:48:46 instance-d java[7058]:         at java.base/sun.security.ssl.SSLSocketInputRecord.handleUnknownRecord(SSLSocketInputRecord.java:451)                                                              
 Apr 16 02:48:46 instance-d java[7058]:         at java.base/sun.security.ssl.SSLSocketInputRecord.decode(SSLSocketInputRecord.java:175)     
 Apr 16 02:48:46 instance-d java[7058]:         at java.base/sun.security.ssl.SSLTransport.decode(SSLTransport.java:111)                     
 Apr 16 02:48:46 instance-d java[7058]:         at java.base/sun.security.ssl.SSLSocketImpl.decode(SSLSocketImpl.java:1505)                  
 Apr 16 02:48:46 instance-d java[7058]:         at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1420)    
  Apr 16 02:48:46 instance-d java[7058]:         at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:455)           
  Apr 16 02:48:46 instance-d java[7058]:         at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:426)           
  Apr 16 02:48:46 instance-d java[7058]:         at com.sun.mail.util.SocketFetcher.configureSSLSocket(SocketFetcher.java:602)                
 Apr 16 02:48:46 instance-d java[7058]:         at com.sun.mail.util.SocketFetcher.createSocket(SocketFetcher.java:376)                      
 Apr 16 02:48:46 instance-d java[7058]:         at com.sun.mail.util.SocketFetcher.getSocket(SocketFetcher.java:214)                         
 Apr 16 02:48:46 instance-d java[7058]:         at com.sun.mail.smtp.SMTPTransport.openServer(SMTPTransport.java:2160)

```


### 附加信息

_No response_

--- a/src/main/java/run/halo/app/mail/AbstractMailService.java
+++ b/src/main/java/run/halo/app/mail/AbstractMailService.java
@@ -213,12 +213,7 @@ public abstract class AbstractMailService implements MailService {
                 optionService.getByPropertyOrDefault(EmailProperties.PASSWORD, String.class));
             mailProperties.setProtocol(
                 optionService.getByPropertyOrDefault(EmailProperties.PROTOCOL, String.class));
-            // check whether starttls is required to be enabled
-            String serverCheck =
-                optionService.getByPropertyOrDefault(EmailProperties.HOST, String.class);
-            if (serverCheck.contains("outlook.com") || serverCheck.contains("office365.com")
-                || serverCheck.contains("live.com") || serverCheck.contains("mail.me.com")
-                || serverCheck.contains("hotmail.com")) {
+            if (optionService.getByPropertyOrDefault(EmailProperties.STARTTLS, Boolean.class)) {
                 Map<String, String> starttls = new HashMap<>();
                 starttls.put("mail.smtp.starttls.enable", "true");
                 starttls.put("mail.smtp.auth", "true");

--- a/src/main/java/run/halo/app/mail/AbstractMailService.java
+++ b/src/main/java/run/halo/app/mail/AbstractMailService.java
@@ -2,6 +2,8 @@ package run.halo.app.mail;
 
 import java.io.UnsupportedEncodingException;
 import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.function.Consumer;
@@ -211,6 +213,17 @@ public abstract class AbstractMailService implements MailService {
                 optionService.getByPropertyOrDefault(EmailProperties.PASSWORD, String.class));
             mailProperties.setProtocol(
                 optionService.getByPropertyOrDefault(EmailProperties.PROTOCOL, String.class));
+            // check whether starttls is required to be enabled
+            String serverCheck =
+                optionService.getByPropertyOrDefault(EmailProperties.HOST, String.class);
+            if (serverCheck.contains("outlook.com") || serverCheck.contains("office365.com")
+                || serverCheck.contains("live.com") || serverCheck.contains("mail.me.com")
+                || serverCheck.contains("hotmail.com")) {
+                Map<String, String> starttls = new HashMap<>();
+                starttls.put("mail.smtp.starttls.enable", "true");
+                starttls.put("mail.smtp.auth", "true");
+                mailProperties.setProperties(starttls);
+            }
             this.cachedMailProperties = mailProperties;
         }
 

--- a/src/main/java/run/halo/app/model/properties/EmailProperties.java
+++ b/src/main/java/run/halo/app/model/properties/EmailProperties.java
@@ -19,6 +19,11 @@ public enum EmailProperties implements PropertyEnum {
     PROTOCOL("email_protocol", String.class, "smtp"),
 
     /**
+     * Is starttls enabled
+     */
+    STARTTLS("email_starttls", Boolean.class, "false"),
+
+    /**
      * SSL port
      */
     SSL_PORT("email_ssl_port", Integer.class, "465"),

